File: Top_100_Liked_Questions/34._Find_First_and_Last_Position_of_Element_in_Sorted_Array/solution.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        # find left and then right with iteration
        if not nums:
            return [-1,-1]
        
        left = None
        l, r = 0, len(nums)-1
        while True:
            if l+1 >= r:
                if nums[l] == target:
                    left = l
                elif nums[r] == target:
                    left = r
                else:
                    left = -1
                break
            
            m = (l+r)//2
            if nums[m] >= target:
                r = m
            else:
                l = m
        
        if left == -1:
            return [-1, -1]
        
        right = None
        l, r = left, len(nums)-1
        while True:
            if l+1 >= r:
                right = r if nums[r] == target else l
                break
            
            m = (l+r)//2
            if nums[m] <= target:
                l = m
            else:
                r = m
        
        return [left, right]
    
    
    def searchRange2(self, nums: List[int], target: int) -> List[int]:
        # find left and then right with recursion
        if not nums:
            return [-1,-1]
        
        def getLeft(l,r,t):
            if l+1 >= r:
                if nums[l] == t:
                    return l
                elif nums[r] == t:
                    return r
                else:
                    return -1
            
            m = (l+r)//2
            if nums[m] >= t:
                return getLeft(l,m,t)
            else:
                return getLeft(m,r,t)
        
        def getRight(l,r,t):
            if l+1 >= r:
                return r if nums[r] == t else l
            
            m = (l+r)//2
            if nums[m] <= t:
                return getRight(m,r,t)
            else:
                return getRight(l,m,t)
            
        l = getLeft(0, len(nums)-1, target)
        r = -1 if l == -1 else getRight(l, len(nums)-1, target)
        
        return [l,r]

File: Top_100_Liked_Questions/34._Find_First_and_Last_Position_of_Element_in_Sorted_Array/solution.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
    let l = 0;
    let r = nums.length-1;
    while (l < r) {
        let m = (l+r) >>> 1;
        if (nums[m] < target) {
            l = m+1
        } else if (nums[m] > target) {
            r = m-1;
        } else {
            r = m;
        }
    }
    
    if (nums[l] != target) {
        return [-1,-1];
    }
    let left = l;
    
    r = nums.length-1;
    while (l < r) {
        let m = (l+r+1) >>> 1;
        if (nums[m] > target) {
            r = m-1;
        } else {
            l = m;
        }
    }
    
    return [left,r];
}


var searchRange2 = function(nums, target) {
    let findLeft = function(l,r,t) {
        if (l === r) {
            return nums[l] === t ? l : -1;
        }
        
        let m = (l+r) >>> 1;
        if (nums[m] < t) {
            return findLeft(m+1,r,t);
        } else if (nums[m] > t) {
            return findLeft(l,m-1,t);
        } else {
            return findLeft(l,m,t);
        }
    }
    
    let findRight = function(l,r,t) {
        if (l === r) {
            return nums[l] === t ? l : -1;
        }
        
        let m = (l+r+1) >>> 1;
        if (nums[m] > t) {
            return findRight(l,m-1,t);
        } else if (nums[m] < t) {
            return findRight(m+1,r,t);
        } else {
            return findRight(m,r,t);
        }
    }
    
    let l = findLeft(0, nums.length-1, target);
    return l === -1 ? [-1,-1] : [l, findRight(l, nums.length-1, target)];
};